test/metabase/driver/sql_jdbc/connection_test.clj

```diff
@@ -212,3 +221,145 @@
+  (let [args   ["-tcp" "-tcpPort", (str port), "-tcpAllowOthers" "-tcpDaemon"]
+        server (Server/createTcpServer (into-array args))]
+    (doto server (.start))))
+
+(deftest test-ssh-tunnel-connection
+  ;; sqlite cannot be behind a tunnel, h2 is tested below, unsure why others fail
+  (mt/test-drivers (disj (sql-jdbc.tu/sql-jdbc-drivers) :sqlite :h2 :oracle :vertica :presto-jdbc :bigquery-cloud-sdk :redshift :athena)
+    (testing "ssh tunnel is established"
+      (let [tunnel-db-details (assoc (:details (mt/db))
+                                     :tunnel-enabled true
+                                     :tunnel-host "localhost"
+                                     :tunnel-auth-option "password"
+                                     :tunnel-port ssh-test/ssh-mock-server-with-password-port
+                                     :tunnel-user ssh-test/ssh-username
+                                     :tunnel-pass ssh-test/ssh-password)]
+        (t2.with-temp/with-temp [Database tunneled-db {:engine (tx/driver), :details tunnel-db-details}]
+          (mt/with-db tunneled-db
+            (sync/sync-database! (mt/db))
+            (is (= [["Polo Lounge"]]
+                   (mt/rows (mt/run-mbql-query venues {:filter [:= $id 60] :fields [$name]}))))))))))
+
+(deftest test-ssh-tunnel-reconnection
+  ;; for now, run against Postgres, although in theory it could run against many different kinds
+  (mt/test-drivers #{:postgres :mysql}
+    (testing "ssh tunnel is reestablished if it becomes closed, so subsequent queries still succeed"
+      (let [tunnel-db-details (assoc (:details (mt/db))
+                                     :tunnel-enabled true
+                                     :tunnel-host "localhost"
+                                     :tunnel-auth-option "password"
+                                     :tunnel-port ssh-test/ssh-mock-server-with-password-port
+                                     :tunnel-user ssh-test/ssh-username
+                                     :tunnel-pass ssh-test/ssh-password)]
+        (t2.with-temp/with-temp [Database tunneled-db {:engine (tx/driver), :details tunnel-db-details}]
+          (mt/with-db tunneled-db
+            (sync/sync-database! (mt/db))
+            (letfn [(check-row []
+                      (is (= [["Polo Lounge"]]
+                             (mt/rows (mt/run-mbql-query venues {:filter [:= $id 60] :fields [$name]})))))]
+              ;; check that some data can be queried
+              (check-row)
+              ;; kill the ssh tunnel; fortunately, we have an existing function that can do that
+              (ssh/close-tunnel! (sql-jdbc.conn/db->pooled-connection-spec (mt/db)))
+              ;; check the query again; the tunnel should have been reestablished
+              (check-row))))))))
+
+(deftest test-ssh-tunnel-connection-h2
+  (testing (str "We need a customized version of this test for H2. It will bring up a new H2 TCP server, pointing to "
```

Comment:
I see this has been moved, still it might make sense to explain why H2 needs special treatment (not running a TCP server).